packages/SwingSet/src/kernel/state/kernelKeeper.js

```diff
@@ -355,6 +366,67 @@ export default function makeKernelKeeper(storage) {
     return storage.get(k);
   }
 
+  const deadKernelPromises = new Set();
+
+  /**
+   * Increment the reference count associated with some kernel promise.
+   *
+   * @param kernelSlot  The kernel slot whose refcount is to be incremented.
+   *
+   * @throws if `kernelSlot` does not refer to a promise.
+   */
+  function incrementKernelPromiseRefCount(kernelSlot, tag) {
+    if (kernelSlot && parseKernelSlot(kernelSlot).type === 'promise') {
+      const refCount = Nat(Number(storage.get(`${kernelSlot}.refCount`))) + 1;
+      kdebug(`++ ${kernelSlot}  ${tag} ${refCount}`);
+      storage.set(`${kernelSlot}.refCount`, `${refCount}`);
+    }
+  }
+
+  /**
+   * Decrement the reference count associated with some kernel promise.
+   *
+   * @param kernelSlot  The kernel slot whose refcount is to be decremented.
+   *
+   * @return true if the reference count has been decremented to zero, false if it is still non-zero
+   *
+   * @throws if `kernelSlot` does not refer to a promise.
+   * @throws if this tries to decrement the reference count below zero.
+   */
+  function decrementKernelPromiseRefCount(kernelSlot, tag) {
+    if (kernelSlot && parseKernelSlot(kernelSlot).type === 'promise') {
+      let refCount = Nat(Number(storage.get(`${kernelSlot}.refCount`)));
+      assert(refCount > 0, details`refCount underflow {kernelSlot} ${tag}`);
+      refCount -= 1;
+      kdebug(`-- ${kernelSlot}  ${tag} ${refCount}`);
+      storage.set(`${kernelSlot}.refCount`, `${refCount}`);
+      if (refCount === 0) {
+        deadKernelPromises.add(kernelSlot);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  function purgeDeadKernelPromises() {
+    if (enableKernelPromiseGC) {
+      for (const kpid of deadKernelPromises.values()) {
```

Comment:
JS question: is this stable (and thorough) if the `Set` changes during iteration? Recursive GC will trigger that.